main.go

```diff
@@ -1,452 +1,107 @@
-	opts := &command.Opts{
-		Stdin:  strings.NewReader("n"),
-		Stdout: bufio.NewWriter(&outBuffer),
-		Stderr: os.Stderr,
-		Dir:    workDir,
-	}
-	var cmd command.Command
-	if useBundler {
-		cmd = factory.CreateBundleExec(name, args, bundlerVersion.Version, opts)
-	} else {
-		cmd = factory.Create(name, args, opts)
+	dependenciesOpts := EnsureDependenciesOpts{
+		GemVersions:    config.GemVersions,
+		UseBundler:     config.GemVersions.fastlane.Found,
+		WorkDir:        config.WorkDir,
+		UpdateFastlane: config.UpdateFastlane,
 	}
 
-	log.Debugf("$ %s", cmd.PrintableCommandArgs())
-	if err := cmd.Run(); err != nil {
-		if errorutil.IsExitStatusError(err) {
-			return "", fmt.Errorf("fastlane command (%s) failed, output: %s", cmd.PrintableCommandArgs(), outBuffer.String())
-		}
-		return "", fmt.Errorf("fastlane command (%s) failed: %v", cmd.PrintableCommandArgs(), err)
+	if err = buildStep.InstallDependencies(dependenciesOpts); err != nil {
+		buildStep.logger.Println()
+		buildStep.logger.Errorf(formattedError(fmt.Errorf("Failed to install Step dependencies: %w", err)))
+		return 1
 	}
 
-	return outBuffer.String(), nil
-}
-
-func functionName(i interface{}) string {
-	return runtime.FuncForPC(reflect.ValueOf(i).Pointer()).Name()
-}
-
-const notConnected = `Connected Apple Developer Portal Account not found.
-Most likely because there is no Apple Developer Portal Account connected to the build.
-Read more: https://devcenter.bitrise.io/getting-started/configuring-bitrise-steps-that-require-apple-developer-account-data/`
-
-func handleSessionDataError(err error) {
-	if err == nil {
-		return
-	}
-
-	if networkErr, ok := err.(devportalservice.NetworkError); ok && networkErr.Status == http.StatusUnauthorized {
-		fmt.Println()
-		log.Warnf("%s", "Unauthorized to query Connected Apple Developer Portal Account. This happens by design, with a public app's PR build, to protect secrets.")
-
-		return
-	}
-
-	fmt.Println()
-	log.Errorf("Failed to activate Bitrise Apple Developer Portal connection: %s", err)
-	log.Warnf("Read more: https://devcenter.bitrise.io/getting-started/configuring-bitrise-steps-that-require-apple-developer-account-data/")
-}
-
-func main() {
-	var config Config
-	parser := stepconf.NewDefaultEnvParser()
-	if err := parser.Parse(&config); err != nil {
-		failf("Issue with input: %s", err)
-	}
-
-	stepconf.Print(config)
-	log.SetEnableDebugLog(config.VerboseLog)
-	fmt.Println()
-
-	// Validate inputs
-	authInputs := appleauth.Inputs{
-		Username:            config.AppleID,
-		Password:            string(config.Password),
-		AppSpecificPassword: string(config.AppSpecificPassword),
-		APIIssuer:           config.APIIssuer,
-		APIKeyPath:          string(config.APIKeyPath),
-	}
-	if err := authInputs.Validate(); err != nil {
-		failf("Issue with authentication related inputs: %v", err)
-	}
-	authSources, err := parseAuthSources(config.BitriseConnection)
+	runOpts := createRunOptions(config)
+	err = buildStep.Run(runOpts)
```

Comment:
this could be moved to the if statement below this line

Reply:
397058eb3330e463673261015c17e5d4365a9e7c